R/subset_datarqst.R

```diff
@@ -30,73 +42,189 @@
 #' subset_datarqst(Species = c("PARMAJ", "FICALB"))
 #'
 #' #Create a specific subset of species and population combinations
-#' subset_datarqst(filter = c("HOG_PARMAJ", "VEL_FICALB", "VLI_PARMAJ"))
+#' subset_datarqst(PopSpec_Filter = c("HOG_PARMAJ", "VEL_FICALB", "VLI_PARMAJ"))
 #'
 #' }
 
 subset_datarqst <- function(file = file.choose(),
                             PopID = unique(pop_names$code),
                             Species = unique(Species_codes$Code),
-                            filter = NULL){
+                            PopSpec_Filter = NULL,
+                            include_conflicting = FALSE,
+                            output_type = "R",
+                            save = TRUE,
+                            test = FALSE){
+
+  if(!test){
+    message(crayon::bold(crayon::green("Select standard data file (RDS format).")))
+    standard_data <- readRDS(file = file)
+  }else{
+    message(crayon::green("Using 'pipeline_output' as standard data."))
+    standard_data <- pipeline_output
+  }
 
-  standard_data <- readRDS(file = file)
 
-  if(!is.null(filter)){
+  if(!is.null(PopSpec_Filter)){
 
-    unique_pops <- unique(stringr::str_split(string = filter, pattern = "_")[[1]])
+    message(crayon::green("Filtering data using 'PopSpec_Filter'."))
 
-    output_brood <- standard_data$Brood_data %>%
-      dplyr::mutate(Pop_sp = paste(PopID, Species, sep = "_")) %>%
-      dplyr::filter(Pop_sp %in% filter)
+    PopSpec_Filter_original <- PopSpec_Filter
 
-    output_capture <- standard_data$Capture_data %>%
-      dplyr::mutate(Pop_sp = paste(CapturePopID, Species, sep = "_")) %>%
-      dplyr::filter(Pop_sp %in% filter)
+    unique_pops <- unique(stringr::str_split(string = PopSpec_Filter, pattern = "_", simplify = TRUE)[,1])
+
+    if(include_conflicting){
+      #PopSpec_Filter <- c(PopSpec_Filter, paste0(unique_pops, '_CCCCCC')) # Use after pipelines updated to standard format 1.1
+      PopSpec_Filter <- c(PopSpec_Filter, paste0(unique_pops, '_CCCCCC'), paste0(unique_pops, '_CONFLICTED'))
+    }
+
+    output_brood <- standard_data$Brood_data %>%
+      dplyr::mutate(Pop_sp = paste(.data$PopID, .data$Species, sep = "_")) %>%
+      dplyr::filter(.data$Pop_sp %in% PopSpec_Filter) %>%
+      dplyr::select(-.data$Pop_sp)
 
     output_individual <- standard_data$Individual_data %>%
-      dplyr::mutate(Pop_sp = paste(PopID, Species, sep = "_")) %>%
-      dplyr::filter(Pop_sp %in% filter)
+      dplyr::mutate(Pop_sp = paste(.data$PopID, .data$Species, sep = "_")) %>%
+      dplyr::filter(.data$Pop_sp %in% PopSpec_Filter) %>%
+      dplyr::select(-.data$Pop_sp)
+
+    output_capture <- standard_data$Capture_data %>%
+      dplyr::filter(.data$IndvID %in% output_individual$IndvID)
```

Comment:
I've ran the code again on HOG and this time it took 47 minutes. 
I took a closer look at the function and it seemed that this filter (and the filter on l. 112-113) is the time-consuming part. Then I was wondering, it is possible to first apply a PopID filter and then start cycling through the IndvIDs? I applied the filter below and it only took 8 minutes. Could this be a solution to the speed problem or am I missing something?

```suggestion
    output_capture <- standard_data$Capture_data %>%
      dplyr::filter(.data$CapturePopID %in% {{PopID}} & .data$ReleasePopID %in% {{PopID}}) %>%
      dplyr::filter(.data$IndvID %in% output_individual$IndvID)
```

Reply:
Another thought: maybe we should also _group by_ PopID, as IndvIDs might not be unique across populations (especially of different species).

Reply:
This is definitely an issue worth considering. We don't require IndvIDs to be unique across populations so we shouldn't assume this will be the case.

Reply:
@StefanVriend as also discussed in person, yes if this helps with speed then it makes sense to include it. Have added it now (with some modifications that take into account filtering by `PopSpec`vs. filtering by `PopID`and `Species`.

Reply:
@StefanVriend and @LiamDBailey, good point about IndvID potentially not being unique across populations. 

I figured there were 2 ways to address this: 1) make sure IndvID is unique by ensuring this at the pipeline level or 2) take into account that IndvID is linked to PopID when subsetting. 
The problem with 1) is that this will not be able to correctly deal with migrants and transplanted individuals. 

Therefore, I went with 2). Stefan suggested grouping by `PopID` while subsetting by `IndvID`, but I could not think of a good way to do that. Instead, I introduce a temporary variable `PopID_IndvID` and use this during the subsetting process. This may not be the most elegant solution, but it works nicely.